drivers/disk/emmc_stm32.c

```diff
@@ -0,0 +1,437 @@
+		return err;
+	}
+
+	err = HAL_MMC_Init(&priv->hsd);
+	if (err != HAL_OK) {
+		LOG_ERR("failed to init stm32_sdmmc");
+		return -EIO;
+	}
+
+	priv->status = DISK_STATUS_OK;
+	return 0;
+}
+
+static void stm32_sdmmc_access_deinit(struct stm32_sdmmc_priv *priv)
+{
+	HAL_MMC_DeInit(&priv->hsd);
+	stm32_sdmmc_clock_disable(priv);
+}
+
+static int stm32_emmc_access_status(struct disk_info *disk)
+{
+	const struct device *dev = disk->dev;
+	struct stm32_sdmmc_priv *priv = dev->data;
+
+	return priv->status;
+}
+
+static int stm32_emmc_access_read(struct disk_info *disk, uint8_t *data_buf,
+				  uint32_t start_sector, uint32_t num_sector)
+{
+	const struct device *dev = disk->dev;
+	struct stm32_sdmmc_priv *priv = dev->data;
+	int err;
+
+	err = HAL_MMC_ReadBlocks(&priv->hsd, data_buf, start_sector,
+				 num_sector, 30000);
+	if (err != HAL_OK) {
+		LOG_ERR("sd read block failed %d", err);
+		return -EIO;
+	}
+
+	while (HAL_MMC_GetCardState(&priv->hsd) != HAL_SD_CARD_TRANSFER)
+		;
+
+	return 0;
+}
+
+static int stm32_emmc_access_write(struct disk_info *disk,
+				    const uint8_t *data_buf,
+				   uint32_t start_sector, uint32_t num_sector)
+{
+	const struct device *dev = disk->dev;
+	struct stm32_sdmmc_priv *priv = dev->data;
+	int err;
+
+	err = HAL_MMC_WriteBlocks(&priv->hsd, (uint8_t *)data_buf, start_sector,
+				 num_sector, 30000);
+	if (err != HAL_OK) {
+		LOG_ERR("sd write block failed %d", err);
+		return -EIO;
+	}
+	while (HAL_MMC_GetCardState(&priv->hsd) != HAL_SD_CARD_TRANSFER)
+		;
+
+	return 0;
+}
+
+static int stm32_emmc_access_ioctl(struct disk_info *disk, uint8_t cmd,
+				    void *buff)
+{
+	const struct device *dev = disk->dev;
+	struct stm32_sdmmc_priv *priv = dev->data;
+	HAL_MMC_CardInfoTypeDef info;
+	int err;
+
+	switch (cmd) {
+	case DISK_IOCTL_GET_SECTOR_COUNT:
+		err = HAL_MMC_GetCardInfo(&priv->hsd, &info);
+		if (err != HAL_OK) {
+			return -EIO;
+		}
+		*(uint32_t *)buff = info.LogBlockNbr;
+		break;
+	case DISK_IOCTL_GET_SECTOR_SIZE:
+		err = HAL_MMC_GetCardInfo(&priv->hsd, &info);
+		if (err != HAL_OK) {
+			return -EIO;
+		}
+		*(uint32_t *)buff = info.LogBlockSize;
+		break;
+	case DISK_IOCTL_GET_ERASE_BLOCK_SZ:
+		*(uint32_t *)buff = 1;
+		break;
+	case DISK_IOCTL_CTRL_SYNC:
+		/* we use a blocking API, so nothing to do for sync */
+		break;
+	default:
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static const struct disk_operations stm32_emmc_ops = {
+	.init = stm32_emmc_access_init,
+	.status = stm32_emmc_access_status,
+	.read = stm32_emmc_access_read,
+	.write = stm32_emmc_access_write,
+	.ioctl = stm32_emmc_access_ioctl,
+};
+
+static struct disk_info stm32_sdmmc_info = {
+	.name = CONFIG_SDMMC_VOLUME_NAME,
+	.ops = &stm32_emmc_ops,
+};
+
+/*
+ * Check if the card is present or not. If no card detect gpio is set, assume
+ * the card is present. If reading the gpio fails for some reason, assume the
+ * card is there.
+ */
+static bool stm32_sdmmc_card_present(struct stm32_sdmmc_priv *priv)
```

Comment:
sorry, my comment got lost somehow. Most of the code below related to card-present and power is equal to sdmmc_stm32 code, can you please move it from sdmmc_stm32.c to a common file like mmc_stm32_common.c and then reuse in new shim driver?

Reply:
Thinking about it, eMMC never uses card-detect, and the power gpio is better renamed reset. I can simply push a change that removes all the card detect logic, and renames `power-gpios` to `reset-gpios`, but splitting out the common code from the original SD driver runs a risk that I do not have access to any SD cards to be able to test that I did not bread the originall driver, though I agree that ultimately this is a better approach.

Please advise on how you would like to see this resolved.

Reply:
@nickstoughton, we'll be able to test that there is no regression on original SD driver, so you can go ahead as this makes sense.